enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,358 @@
+item is removed, the CVO will apply the previously filtered(skipped) resources to the cluster on the next reconciliation.
+Adding an item to the filtered list is not supported because it requires the component be removed from the
+running cluster which has more significant implications for how all traces of the component are removed.
+
+The currently configured filter list for the CVO should be recorded in telemeter so we can understand
+the configuration of a given cluster.
+
+If we want components to be disabled by default (either existing components or new ones added in the future),
+we can add their component names to a list of default-disabled components that the installer populates
+the install-config with.  Users can then edit that disabled list in the install-config to remove those
+components if they want them enabled, and add other components they'd like to disable.
+
+In the future, we might allow specific APIs to be disabled, such as the build api.  This could be done by
+defining additional capability keywords that can be put into the install-config field being defined here,
+which would drive the creation of cluster config that the apiserver+controllers used to disable that particular
+api.
+
+
+### Risks and Mitigations
+
+The primary risk is that teams understand how to use these new annotations and apply them
+correctly to the full set of resources that make up their component.  Inconsistent or
+partial labeling will result in inconsistent or partially deployed resources for a component.
+
+Another risk is that this introduces more deployment configurations which might
+have unforeseen consequences (e.g. not installing the imageregistry causes some
+other component that assumes there is always an imageregistry or assumes the
+presence of some CRD api that is installed with the imageregistry to break).
+
+There was some discussion about the pros/cons of allowing each component to be enabled/disabled independent
+of that component explicitly opting into a particular (presumably well tested) configuration/topology
+[here](https://github.com/openshift/enhancements/pull/200#discussion_r375837903).  The position of this EP is that
+we should only recommend the exclusion of fully independent "capability" components that are not depended on by
+other components.  Further the assumption is that it will be reasonable to tell a customer who disabled
+something and ended up with a non-functional cluster that their chosen exclusions are simply not supported
+currently, or that they must accept the degraded functionality caused by the missing dependency, if they
+intend to keep it disabled.
+
+Since the only components/resources that can be filtered out of the installation are ones that are explicilty
+annotated with `capability.openshift.io/$component`, end-users will not be able to use this mechanism to filter
+components/resources that we did not intend for them to be able to filter out.
```

Comment:
the CVO could put that in status as a warning.  I don't think the installer can know at all (w/o cracking open a payload and seeing what's in there).

but ultimately if you specified a keyword to filter, and it didn't match anything, i'm reluctant to call that an "error" (maybe you put it in there knowing that when you upgrade to 4.12, it will match a new component that's coming in that version, that you do not want).

Reply:
(will add wording to this effect)

Reply:
> ...w/o cracking open a payload and seeing what's in there...

I'd floated installer access to the release image way back in openshift/installer#1286.  Skimming that discussion again, I don't see a good comment to link, but I recall one of the issues being discussed is:

* Installer has access to the provider API, but
* Installer does not have access to the install-serving registry

which might happen when installing into restricted networks.  There have also been a number of issues with `oc` failing to function as expected in restricted networks, [because it was unaware of cluster-side mirror configuration][1].

However, the set of optional components may be small and stable enough that the installer could hard-code the validation logic locally, and then still perform validation based on it's hard-coded guesses about how the associated release was set up.  Not ideal from a maintenance perspective, but not impossible.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1823143

Reply:
@wking This seems different than that because it doesn't vary based on other values. So the installer, at build time, could be made aware of valid capabilities as a byproduct of the release image being an input to that process.